File: src/utils/categoryIcon.ts
import { Colors } from '@/src/styles/theme';

interface IconSpec {
    name: string;
    color: string;
}

type CategoryType = 'expense' | 'income';

const NORMALIZE = (s: string) => (s || '').toLowerCase().trim();

const MAP: Record<string, IconSpec> = {
    groceries:    { name: 'category_groceries',  color: Colors.cat.groceries },
    grocery:      { name: 'category_groceries',  color: Colors.cat.groceries },
    food:         { name: 'category_food',       color: Colors.warning },
    dining:       { name: 'category_food',       color: Colors.warning },
    restaurant:   { name: 'category_food',       color: Colors.warning },
    restaurants:  { name: 'category_food',       color: Colors.warning },
    travel:       { name: 'category_travel',     color: Colors.cat.travel },
    transport:    { name: 'category_transport',  color: Colors.cat.car },
    transportation: { name: 'category_transport', color: Colors.cat.car },
    car:          { name: 'category_car',        color: Colors.cat.car },
    fuel:         { name: 'category_fuel',       color: Colors.warning },
    gas:          { name: 'category_gas',        color: Colors.warning },
    petrol:       { name: 'category_fuel',       color: Colors.warning },
    home:         { name: 'category_home',       color: Colors.cat.home },
    rent:         { name: 'category_rent',       color: Colors.cat.rent },
    bills:        { name: 'category_bills',      color: Colors.deepPurple },
    bill:         { name: 'category_bills',      color: Colors.deepPurple },
    utilities:    { name: 'category_bills',      color: Colors.deepPurple },
    electricity:  { name: 'category_electricity', color: Colors.warning },
    insurance:    { name: 'category_insurance',  color: Colors.cat.insurance },
    insurances:   { name: 'category_insurance',  color: Colors.cat.insurance },
    education:    { name: 'category_education',  color: Colors.cat.education },
    marketing:    { name: 'category_marketing',  color: Colors.cat.marketing },
    shopping:     { name: 'category_shopping',   color: Colors.cat.shopping },
    internet:     { name: 'category_internet',   color: Colors.cat.internet },
    mobile:       { name: 'category_mobile',     color: Colors.blue },
    phone:        { name: 'category_mobile',     color: Colors.blue },
    healthcare:   { name: 'category_healthcare', color: Colors.negative },
    health:       { name: 'category_healthcare', color: Colors.negative },
    medical:      { name: 'category_healthcare', color: Colors.negative },
    entertainment:{ name: 'category_entertainment', color: Colors.pink },
    water:        { name: 'category_water',      color: Colors.cat.water },
    gym:          { name: 'category_gym',        color: Colors.cat.gym },
    subscription: { name: 'category_subscription', color: Colors.cat.subscription },
    subscriptions:{ name: 'category_subscription', color: Colors.cat.subscription },
    gifts:        { name: 'category_gifts',      color: Colors.warning },
    gift:         { name: 'category_gifts',      color: Colors.warning },
    vacation:     { name: 'category_vacation',   color: Colors.cat.vacation },
    salary:       { name: 'type_income',         color: Colors.primary },
    bonus:        { name: 'income_bonus',        color: Colors.warning },
    freelance:    { name: 'income_freelance',    color: Colors.cyan },
    investment:   { name: 'income_investment',   color: Colors.accentDark },
    other:        { name: 'income_other',        color: Colors.cat.other },
};

const INCOME_MAP: Record<string, IconSpec> = {
    salary:       { name: 'type_income',         color: Colors.primary },
    bonus:        { name: 'income_bonus',        color: Colors.warning },
    freelance:    { name: 'income_freelance',    color: Colors.cyan },
    business:     { name: 'income_business',     color: Colors.primary },
    sales:        { name: 'income_business',     color: Colors.primary },
    commission:   { name: 'income_commission',   color: Colors.accentDark },
    tips:         { name: 'earn_rewards',        color: Colors.warning },
    investment:   { name: 'income_investment',   color: Colors.accentDark },
    'rental income': { name: 'income_rental',    color: Colors.cat.rent },
    rent:         { name: 'income_rental',       color: Colors.cat.rent },
    interest:     { name: 'income_interest',     color: Colors.accentDark },
    dividends:    { name: 'income_investment',   color: Colors.accentDark },
    refund:       { name: 'income_refund',       color: Colors.primary },
    reimbursement:{ name: 'income_refund',       color: Colors.primary },
    gift:         { name: 'income_gift',         color: Colors.warning },
    cashback:     { name: 'earn_rewards',        color: Colors.warning },
    other:        { name: 'income_other',        color: Colors.primary },
};

const DEFAULT: IconSpec = { name: 'category_other', color: Colors.primary };

export function resolveCategoryIcon(category?: string, type?: CategoryType): IconSpec {
    if (!category) return DEFAULT;
    const key = NORMALIZE(category);
    if (type === 'income') return INCOME_MAP[key] || MAP[key] || DEFAULT;
    return MAP[key] || DEFAULT;
}

File: lib/core/utils/category_icon.dart
import 'package:flutter/material.dart';

import '../theme/app_colors.dart';

/// Resolved icon for a category: a WealthifyIcon glyph name + a brand color.
class CategoryIconSpec {
  const CategoryIconSpec(this.name, this.color);
  final String name;
  final Color color;
}

/// Maps a color key (as used in the RN `categoryIcon.ts`) onto the active palette.
Color _pick(AppColors c, String key) {
  switch (key) {
    case 'warning':
      return c.warning;
    case 'deepPurple':
      return c.deepPurple;
    case 'blue':
      return c.blue;
    case 'negative':
      return c.negative;
    case 'pink':
      return c.pink;
    case 'cyan':
      return c.cyan;
    case 'accentDark':
      return c.accentDark;
    case 'primary':
      return c.primary;
  }
  if (key.startsWith('cat.')) return c.categoryColor(key.substring(4));
  return c.primary;
}

// (glyphName, colorKey) — ported 1:1 from legacy_rn/src/utils/categoryIcon.ts
const Map<String, (String, String)> _expense = {
  'groceries': ('category_groceries', 'cat.groceries'),
  'grocery': ('category_groceries', 'cat.groceries'),
  'food': ('category_food', 'warning'),
  'dining': ('category_food', 'warning'),
  'restaurant': ('category_food', 'warning'),
  'restaurants': ('category_food', 'warning'),
  'travel': ('category_travel', 'cat.travel'),
  'transport': ('category_transport', 'cat.car'),
  'transportation': ('category_transport', 'cat.car'),
  'car': ('category_car', 'cat.car'),
  'fuel': ('category_fuel', 'warning'),
  'gas': ('category_gas', 'warning'),
  'petrol': ('category_fuel', 'warning'),
  'home': ('category_home', 'cat.home'),
  'rent': ('category_rent', 'cat.rent'),
  'bills': ('category_bills', 'deepPurple'),
  'bill': ('category_bills', 'deepPurple'),
  'utilities': ('category_bills', 'deepPurple'),
  'electricity': ('category_electricity', 'warning'),
  'insurance': ('category_insurance', 'cat.insurance'),
  'insurances': ('category_insurance', 'cat.insurance'),
  'education': ('category_education', 'cat.education'),
  'marketing': ('category_marketing', 'cat.marketing'),
  'shopping': ('category_shopping', 'cat.shopping'),
  'internet': ('category_internet', 'cat.internet'),
  'mobile': ('category_mobile', 'blue'),
  'phone': ('category_mobile', 'blue'),
  'healthcare': ('category_healthcare', 'negative'),
  'health': ('category_healthcare', 'negative'),
  'medical': ('category_healthcare', 'negative'),
  'entertainment': ('category_entertainment', 'pink'),
  'water': ('category_water', 'cat.water'),
  'gym': ('category_gym', 'cat.gym'),
  'subscription': ('category_subscription', 'cat.subscription'),
  'subscriptions': ('category_subscription', 'cat.subscription'),
  'gifts': ('category_gifts', 'warning'),
  'gift': ('category_gifts', 'warning'),
  'vacation': ('category_vacation', 'cat.vacation'),
  'salary': ('type_income', 'primary'),
  'bonus': ('income_bonus', 'warning'),
  'freelance': ('income_freelance', 'cyan'),
  'investment': ('income_investment', 'accentDark'),
  'other': ('income_other', 'cat.other'),
};

const Map<String, (String, String)> _income = {
  'salary': ('type_income', 'primary'),
  'bonus': ('income_bonus', 'warning'),
  'freelance': ('income_freelance', 'cyan'),
  'business': ('income_business', 'primary'),
  'sales': ('income_business', 'primary'),
  'commission': ('income_commission', 'accentDark'),
  'tips': ('earn_rewards', 'warning'),
  'investment': ('income_investment', 'accentDark'),
  'rental income': ('income_rental', 'cat.rent'),
  'rent': ('income_rental', 'cat.rent'),
  'interest': ('income_interest', 'accentDark'),
  'dividends': ('income_investment', 'accentDark'),
  'refund': ('income_refund', 'primary'),
  'reimbursement': ('income_refund', 'primary'),
  'gift': ('income_gift', 'warning'),
  'cashback': ('earn_rewards', 'warning'),
  'other': ('income_other', 'primary'),
};

CategoryIconSpec resolveCategoryIcon(
  String? category, {
  bool income = false,
  required AppColors colors,
}) {
  if (category == null || category.trim().isEmpty) {
    return CategoryIconSpec('category_other', colors.primary);
  }
  final key = category.toLowerCase().trim();
  final hit = income ? (_income[key] ?? _expense[key]) : _expense[key];
  final spec = hit ?? ('category_other', 'primary');
  return CategoryIconSpec(spec.$1, _pick(colors, spec.$2));
}
